### ISSUE TITLE
[security issue] Updating an alias via API removes it from other firewall rules that use it, disabling all pf rules in the process

### ISSUE THREAD

**Issue by wrobelda** [opnsense/core#6208]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [x] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Describe the bug**

I use the API programmatically update the IP that the alias points to, like this:
`https://10.0.0.1/api/firewall/alias/setItem/a1e0c761-4236-4869-xxxx-148abcddeeff`

with following payload: `{"alias": { "content": "192.168.1.2" }}`

These calls work as expected and update the alias appropriately. However, after updating the alias, it is wiped from another NAT port-forward rule I use it in.

Normally, the rule reads:
`rdr log on vtnet0 inet proto udp from {10.0.0.104} to {any} port {53} -> $alias_in_speech port 53`

<img width="708" alt="Screenshot 2022-12-21 at 10 07 09 PM" src="https://user-images.githubusercontent.com/1702435/209002972-5184a698-afe9-4d70-a7ad-ab48bed5d887.png">

however, after the call, it changes to:

`rdr log on vtnet0 inet proto udp from {10.0.0.104} to {any} port {53} port 53`

<img width="733" alt="Screenshot 2022-12-21 at 10 07 52 PM" src="https://user-images.githubusercontent.com/1702435/209002995-4cfe988a-64f4-42db-a157-080ba04d81a3.png">

Note that while I can see the alias disappearing from the rule in the UI immediately after the API call, the firewall rule itself  doesn't update until after a reboot.

And this is where things get serious, because since that rule missing the alias is an improper one, it ends up causing an error:

```
There were error(s) loading the rules: /tmp/rules.debug:93: syntax error - The line in question reads [93]: rdr log on vtnet0 inet proto udp from {10.0.0.104} to {any} port {53} port 53
```

This, in turn, disables *all* of the rules altogether, leaving the system completely open:

```
/usr/local/etc/rc.filter_configure: The command '/sbin/pfctl -f /tmp/rules.debug.old' returned exit code '1', the output was '/tmp/rules.debug.old:93: syntax error pfctl: Syntax error in config file: pf rules not loaded'
```

Therefore, I consider this a serious security issue: an actor with limited permissions can trigger a simple API call which can potentially disable all of the rules altogether.

**Expected behavior**
1. Updating alias via API call should not affect other rules
2. One could argue that the rules added/modified by the API should be denoted as such *and* if such rule becomes broken, only that particular rule gets dropped, instead of leaving the system completely open.

**Environment**

Software version used and hardware type if relevant, e.g.:

OPNsense 22.7.9_3-amd64
FreeBSD 13.1-RELEASE-p5
OpenSSL 1.1.1s 1 Nov 2022
Intel(R) Core(TM) i5-10500T CPU @ 2.30GHz (4 cores, 4 threads)

**Comment by wrobelda** [comment]
@AdSchellevis if I understand https://github.com/opnsense/core/commit/f95ddbc695c155b6557829640e3486d917562142 correctly, it fixes applying invalid pf rules. But does it fix the underlying reason why rules that use an alias get updated after said alias is updated?

**Comment by AdSchellevis** [comment]
they aren't.  commit message should be clear..

**Comment by wrobelda** [comment]
Well, I applied this patch using `opnsense-patch f95ddbc` and the behavior is the same: `alias_in_speech` is removed from a NAT rule as soon as I make that alias API call.

**Comment by wrobelda** [comment]
@AdSchellevis would you mind offering some clarity on this issue, please? From my end, it doesn't look like it is fixed.

**Comment by kulikov-a** [comment]
@wrobelda 
are you sure that you made the right payload with `{"alias": { "content": "192.168.1.2" }}` ?
skipping the name may trigger model refactoring and wipe alias "by name" references from rules. imho the name field should always be set 
https://github.com/opnsense/core/blob/e27234b4ccc0f94372467efd76513c8a0ad18413/src/opnsense/mvc/app/controllers/OPNsense/Firewall/Api/AliasController.php#L124
(although the api call can probably be improved a bit in terms of error protection, imho it is always better to note how the gui does it)

> could argue that the rules added/modified by the API should be denoted as such and if such rule becomes broken, only that particular rule gets dropped
the one rule drop can lead to an unpredictable result imho?

**Comment by AdSchellevis** [comment]
@wrobelda I'm not sure what you are doing, but te content of the alias (for all network types) is detached from the rules themselves, which makes this more or less impossible. You already consumed way more community support time than I'm willing to spend to be honest and in my humble opinion are often blaming others for local issues or lack of understanding. 

You're free to open issues and ask others for help,  just don't expect too much from my end anymore. 

I can recommend to read the following book to better understand how opensource projects operate by the way https://www.amazon.com/Working-Public-Making-Maintenance-Software/dp/0578675862

**Comment by wrobelda** [comment]
> @wrobelda are you sure that you made the right payload with `{"alias": { "content": "192.168.1.2" }}` ? skipping the name may trigger model refactoring and wipe alias "by name" references from rules. imho the name field should always be set

Indeed, I added the name and it does work properly now. 


>  imho the name field should always be set
> https://github.com/opnsense/core/blob/e27234b4ccc0f94372467efd76513c8a0ad18413/src/opnsense/mvc/app/controllers/OPNsense/Firewall/Api/AliasController.php#L124

This looks more like a bug to me, though, if I am not mistaken. [This line](https://github.com/opnsense/core/blob/e27234b4ccc0f94372467efd76513c8a0ad18413/src/opnsense/mvc/app/controllers/OPNsense/Firewall/Api/AliasController.php#LL126C42-L126C42):
`            if ($new_name != $old_name) {`

should rather read:

`            if ($new_name !== null && $new_name != $old_name) {`

because setting the name to 'empty' should not be allowed?

> imho it is always better to note how the gui does it)

True, the GUI sets all the fields in the model at once, I assumed it is not necessary and a single field should work, too — which it does, except as you noted, the name is also required.

> > could argue that the rules added/modified by the API should be denoted as such and if such rule becomes broken, only that particular rule gets dropped
>
> the one rule drop can lead to an unpredictable result imho?

My logic here is: since I am using the API and there's no validation like when setting things up in the GUI, if my API call is incorrect I can expect it not to be applied. Dropping _all_ of the Firewall rules at this point is much more severe of a side effect. But I believe this actually issue was already solved.

**Comment by wrobelda** [comment]
FYI, I tested the `if ($new_name !== null && $new_name != $old_name) {` fix and it solved the problem, so I created a PR #6254 with a fix.

**Comment by kulikov-a** [comment]
> and there's no validation like when setting things up in the GUI

Why do you think so?
you didn't change the value to empty, you didn't send the name value at all.
you can try with the error in ip to make sure that validation works

I don't know how you implement using the API, but it seems logical to me that to change an item, the current item is first requested, then the required fields are changed, and then the entire item is sent back. In this case, such errors should not occur?
